syntax = "proto3";

package textdiffusion.v1;

// Minimal finish reasons for text generation.
enum FinishReason {
  FINISH_REASON_LENGTH = 0;
}

// Simple text generation service with synchronous one-shot generation.
service TextGenerationService {
  rpc Generate(GenerateRequest) returns (GenerateResponse);
}

// Tokens abstraction
message Tokens {
  // Token IDs
  repeated uint32 ids = 1;
}

// Request for a single generation run.
message GenerateRequest {
  // Input tokenized prompt to condition the diffusion process.
  Tokens prompt_tokens = 1;

  // Maximum number of output tokens (or characters, depending on server design).
  // If reached, finish_reason = FINISH_REASON_LENGTH.
  int32 max_output_tokens = 2;

  // Number of diffusion or sampling iterations to run.
  int32 num_steps = 3;

  // Random seed.
  uint64 seed = 4;

  // Mask token id.
  uint32 mask_id = 5;

  // Block length.
  int32 block_length = 6;

  // Sampling temperature / noise scale.
  // For masked discrete diffusion you might treat 0.0 as deterministic.
  float temperature = 7;

  // Client-defined correlation ID.
  string request_id = 8;
}

// Response for a single generation run.
message GenerateResponse {
  // The final generated tokens.
  Tokens output_tokens = 1;

  // Why generation stopped.
  FinishReason finish_reason = 2;

  // Echo of request_id, if provided.
  string request_id = 3;
}
